webnn/resources/utils.js

```diff
@@ -1,61 +1,312 @@
+    for (let subInput of resources.inputs) {
+      inputs[subInput.name] = new TypedArrayDict[subInput.type](subInput.data);
+    }
+  } else {
+    for (let inputName in resources.inputs) {
+      const subTestByName = resources.inputs[inputName];
+      inputs[inputName] = new TypedArrayDict[subTestByName.type](subTestByName.data);
+    }
+  }
+  let outputs = {};
+  if (Array.isArray(resources.expected)) {
+    // the outputs of split() or gru() is a sequence
+    for (let i = 0; i < resources.expected.length; i++) {
+      const subExpected = resources.expected[i];
+      outputs[subExpected.name] = new TypedArrayDict[subExpected.type](sizeOfShape(subExpected.shape));
+      namedOperands[subExpected.name] = outputOperand[i];
+    }
+  } else {
+    const outputName = resources.expected.name ? resources.expected.name : 'output';
+    // matmul 1D with 1D produces a scalar which doesn't have its shape
+    const shape = resources.expected.shape ? resources.expected.shape : [];
+    outputs[outputName] = new TypedArrayDict[resources.expected.type](sizeOfShape(shape));
+    namedOperands[outputName] = outputOperand;
+  }
+  return [namedOperands, inputs, outputs];
+};
+
+/**
+ * Build a graph, synchronously compile graph and execute, then check computed results.
+ * @param {String} name - An operation name
+ * @param {MLContext} context - A ML context
+ * @param {MLGraphBuilder} builder - A ML graph builder
+ * @param {Object} resources - Resources used for building a graph
+ * @param {Function} buildFunc - A build function for an operation
+ */
+const runSync = (name, context, builder, resources, buildFunc) => {
+  // build a graph
+  const [namedOutputOperands, inputs, outputs] = buildGraph(builder, resources, buildFunc, name);
+  // synchronously compile the graph up to the output operand
+  const graph = builder.buildSync(namedOutputOperands);
+  // synchronously execute the compiled graph.
+  context.computeSync(graph, inputs, outputs);
+  checkResults(name, namedOutputOperands, outputs, resources);
+};
+
+/**
+ * Build a graph, asynchronously compile graph and execute, then check computed results.
+ * @param {String} name - An operation name
+ * @param {MLContext} context - A ML context
+ * @param {MLGraphBuilder} builder - A ML graph builder
+ * @param {Object} resources - Resources used for building a graph
+ * @param {Function} buildFunc - A build function for an operation
+ */
+const run = async (name, context, builder, resources, buildFunc) => {
+  // build a graph
+  const [namedOutputOperands, inputs, outputs] = buildGraph(builder, resources, buildFunc, name);
+  // asynchronously compile the graph up to the output operand
+  const graph = await builder.build(namedOutputOperands);
+  // asynchronously execute the compiled graph
+  await context.compute(graph, inputs, outputs);
+  checkResults(name, namedOutputOperands, outputs, resources);
+};
+
+/**
+ * Run WebNN operation tests.
+ * @param {String} name - An operation name
```

Comment:
Should this parameter be named `operationName`? `name` is not clear about whose name it is. Please make this one and all the other `name`s to be more specific.

Reply:
Updated, thanks.